Ready for some feedback (#393)
ITU-BDSA23-GROUP12/Chirp: docs/report.md

## Sequence of functionality/calls trough _Chirp!_

# Process
## Build, test, release, and deployment

> During development of the application, we have used github workflows to automate some processes. It is defined in a YAML file and set to run on a trigger event. We have used one for Build and testing the application, one for creating and releasing a single file .NET application to Github and one for deploying our .NET app to Azure and the schemas for our database. Workflows consist of one or more jobs, which have a sequence of steps that has to be executed. The diagrams below illustrate the steps and jobs of each workflow.

Reviewer: ```suggestion
During development of the application, we have used github workflows to automate some processes. It is defined in a YAML file and set to run on a trigger event. We have used one for Build and testing the application, one for creating a release on Github and one for deploying our .NET app to Azure and the schemas for our database. Workflows consist of one or more jobs, which have a sequence of steps that has to be executed. The diagrams below illustrate the steps and jobs of each workflow.
```

### Build and test workflow

> This workflow is triggered when a branch is pushed or a pull request is created. It locates the source code, sets up a .NET Core environment. Then it downloads the missing if any dependencies are missing, builds the application and the test suite is run. 

Reviewer 2: ```suggestion
This workflow is triggered when a branch is pushed or a pull request is created. It locates the source code and sets up a .NET Core environment. Then it downloads dependencies, if any are missing, builds the application and runs the test suite.
```

This workflow is useful streamlining reviewing pull requests. See figure \ref{BuildAndTest-workflow}
![UML Diagram of build and test workflow\lable{BuildAndTest-workflow}](https://github.com/ITU-BDSA23-GROUP12/Chirp/blob/main/docs/images/BuildAndTest-Workflow.drawio.png?raw=true)

### Release workflow
Whenever a new tag is pushed into main with our format, the release workflow is triggered. Again the .NET Core environment is set up, and then we add the artifact to our release with softprops/action-gh-release.

> This streamlining helped making releases with executables on Github, whenever our main was given new tags. We manually had to check if Main built and passed the tests before given new tags, but could have been included. See figure \ref{Release-workflow}

Reviewer: ```suggestion
Initially this streamlining helped making releases with executables on Github, whenever our main was given new tags. However we have later switched to only release source code, as we have evolved _Chirp!_ into a web application. We manually had to check if Main built and passed the tests before given new tags, but could have been included. See figure \ref{Release-workflow}
```

![UML Diagram of the release workflow\lable{Release-workflow}](https://github.com/ITU-BDSA23-GROUP12/Chirp/blob/main/docs/images/Release-Workflow.drawio.png)

### Deployment workflow

> This workflow consist of to jobs: 'BuildAndTest' and 'deploy'. Jobs can be run concurrently, but we need 'BuildAndTest' to run successfully, before we want to bother with deploying, hence the key word 'needs', which only runs the 'deploy' when 'BuildAndTest' is done. 

Reviewer 3: ```suggestion
This workflow consist of to jobs: 'BuildAndTest' and 'deploy'. Jobs can be run concurrently, but we need 'BuildAndTest' to run successfully, before we want to bother with deploying, hence the key word 'needs', which means we only run the 'deploy' when 'BuildAndTest' is done. 
```

> can manually be triggered on github or triggered by push to main. Again it sets up a .Net Core environment, restores the dependencies, builds and runs our test suite. Create and upload application  artifacts to the github actions workflow system with the name '.net-app'.  Then the new migrations are bundled together and uploaded with the name 'efbundle'.

Reviewer 3: Mangler der noget her?

> Then the second part of the workflow called 'deploy', takes care of uploading the new web-app to azure and deploy the new migrations to the azure database. See figure \ref{deploy-workflow}

Reviewer 3: ```suggestion
Then the second part of the workflow called 'deploy', takes care of uploading the new web-app to Azure and deploy the new migrations to the Azure database. See figure \ref{deploy-workflow}
```

> ![UML Diagram of the deploy workflow\lable{deploy-workflow}](https://github.com/ITU-BDSA23-GROUP12/Chirp/blob/main/docs/images/Release-Workflow.drawio.png)

Reviewer 3: Er det et forkert billede?
```suggestion
![UML Diagram of the deploy workflow\lable{deploy-workflow}](https://github.com/ITU-BDSA23-GROUP12/Chirp/blob/main/docs/images/Release-Workflow.drawio.png?raw=true)
```
